README
The cabal-install package
=========================

[Cabal home page](http://www.haskell.org/cabal/)

The `cabal-install` package provides a command line tool called `cabal`. The
tool uses the `Cabal` library and provides a convenient user interface to the
Cabal/Hackage package build and distribution system. It can build and install
both local and remote packages, including dependencies.


Installation instructions for the cabal-install command line tool
=================================================================

The `cabal-install` package requires a number of other packages, most of which
come with a standard ghc installation. It requires the `network` package, which
is sometimes packaged separately by Linux distributions, for example on
debian or ubuntu it is in "libghc6-network-dev".

It requires a few other Haskell packages that are not always installed:

 * Cabal  (version 1.10 or later)
 * HTTP   (version 4000 or later)
 * zlib   (version 0.4  or later)

All of these are available from [Hackage](http://hackage.haskell.org).

Note that on some Unix systems you may need to install an additional zlib
development package using your system package manager, for example on
debian or ubuntu it is in "zlib1g-dev". It is needed is because the
Haskell zlib package uses the system zlib C library and header files.

The `cabal-install` package is now part of the Haskell Platform so you do not
usually need to install it separately. However if you are starting from a
minimal ghc installation then you need to install `cabal-install` manually.
Since it is just an ordinary Cabal package it can be built in the standard
way, but to make it a bit easier we have partly automated the process:


Quickstart on Unix systems
--------------------------

As a convenience for users on Unix systems there is a `bootstrap.sh` script
which will download and install each of the dependencies in turn.

    $ ./bootstrap.sh

It will download and install the above three dependencies. The script will
install the library packages into `$HOME/.cabal/` and the `cabal` program will
be installed into `$HOME/.cabal/bin/`.

You then have two choices:

 * put `$HOME/.cabal/bin` on your `$PATH`
 * move the `cabal` program somewhere that is on your `$PATH`

The next thing to do is to get the latest list of packages with:

    $ cabal update

This will also create a default config file (if it does not already echo exist)
at `$HOME/.cabal/config`

By default cabal will install programs to `$HOME/.cabal/bin`. If you do not
want to add this directory to your `$PATH` then you can change the setting in
the config file, for example you could use:

    symlink-bindir: $HOME/bin


Quickstart on Windows systems
-----------------------------

For Windows users we provide a pre-compiled [cabal.exe] program. Just download
it and put it somewhere on your `%PATH%`, for example
`C:\Program Files\Haskell\bin`.

[cabal.exe]: http://haskell.org/cabal/release/cabal-install-latest/cabal.exe

The next thing to do is to get the latest list of packages with

    cabal update

This will also create a default config file (if it does not already echo exist)
at `C:\Documents and Settings\username\Application Data\cabal\config`


Using cabal-install
===================

There are two sets of commands: commands for working with a local project build
tree and ones for working with distributed released packages from hackage.

For a list of the full set of commands and the flags for each command see

    $ cabal --help


Commands for developers for local build trees
---------------------------------------------

The commands for local project build trees are almost exactly the same as the
`runghc Setup` command line interface that many people are already familiar
with. In particular there are the commands

    cabal configure
    cabal build
    cabal haddock
    cabal clean
    cabal sdist

The `install` command is somewhat different. It is an all-in-one operation. If
you run

    $ cabal install

in your build tree it will configure, build and install. It takes all the flags
that `configure` takes such as `--global` and `--prefix`.

In addition, if any dependencies are not installed it will download and install
them. If can also rebuild packages to ensure a consistent set of dependencies.


Commands for released hackage packages
--------------------------------------

    $ cabal update

This command gets the latest list of packages from the hackage server.
Currently this command has to be run manually occasionally, in particular if
you want to install a newly released package. 


    $ cabal install xmonad

This is the eponymous command. It installs one or more named packages (and all
their dependencies) from hackage.

By default it installs the latest available version however you can optionally
specify exact versions or version ranges. For example `cabal install alex-2.2`
or `cabal install parsec < 3`.

    $ cabal upgrade xmonad

This is a variation on the `install` command. Both mean to install the latest
version, the only difference is in the treatment of dependencies. The `install`
command tries to use existing installed versions of dependent packages while
the `upgrade` command tries to upgrade all the dependencies too.

    $ cabal list xml

This does a search of the installed and available packages. It does a
case-insensitive substring match on the package name.


changelog
-*-change-log-*-

0.10.2 Duncan Coutts <duncan@community.haskell.org> March 2011
	* Fix assertion failure in cabal configure
	* Fix command line handling for "world" target

0.10.0 Duncan Coutts <duncan@community.haskell.org> February 2011
	* New package targets: local dirs, local and remote tarballs
	* Initial support for a "world" package target
	* Partial fix for situation where user packages mask global ones
	* Removed cabal upgrade, new --upgrade-dependencies flag
	* New cabal install --only-dependencies flag
	* New cabal fetch --no-dependencies and --dry-run flags
	* Improved output for cabal info
	* Simpler and faster bash command line completion
	* Fix for broken proxies that decompress wrongly
	* Fix for cabal unpack to preserve executable permissions
	* Adjusted the output for the -v verbosity level in a few places

0.8.2 Duncan Coutts <duncan@community.haskell.org> March 2010
	* Fix for cabal update on Windows
	* On windows switch to per-user installs (rather than global)
	* Handle intra-package dependencies in dependency planning
	* Minor tweaks to cabal init feature
	* Fix various -Wall warnings
	* Fix for cabal sdist --snapshot

0.8.0 Duncan Coutts <duncan@haskell.org> Dec 2009
	* Works with ghc-6.12
	* New "cabal init" command for making initial project .cabal file
	* New feature to maintain an index of haddock documentation

0.6.4 Duncan Coutts <duncan@haskell.org> Nov 2009
	* Improve the algorithm for selecting the base package version
	* Hackage errors now reported by "cabal upload [--check]"
	* Improved format of messages from "cabal check"
	* Config file can now be selected by an env var
	* Updated tar reading/writing code
	* Improve instructions in the README and bootstrap output
	* Fix bootstrap.sh on Solaris 9
	* Fix bootstrap for systems where network uses parsec 3
	* Fix building with ghc-6.6

0.6.2 Duncan Coutts <duncan@haskell.org> Feb 2009
	* The upgrade command has been disabled in this release
	* The configure and install commands now have consistent behaviour
	* Reduce the tendancy to re-install already existing packages
	* The --constraint= flag now works for the install command
	* New --preference= flag for soft constraints / version preferences
	* Improved bootstrap.sh script, smarter and better error checking
	* New cabal info command to display detailed info on packages
	* New cabal unpack command to download and untar a package
	* HTTP-4000 package required, should fix bugs with http proxies
	* Now works with authenticated proxies.
	* On Windows can now override the proxy setting using an env var
	* Fix compatability with config files generated by older versions
	* Warn if the hackage package list is very old
	* More helpful --help output, mention config file and examples
	* Better documentation in ~/.cabal/config file
	* Improved command line interface for logging and build reporting
	* Minor improvements to some messages

0.6.0 Duncan Coutts <duncan@haskell.org> Oct 2008
	* Constraint solver can now cope with base 3 and base 4
	* Allow use of package version preferences from hackage index
	* More detailed output from cabal install --dry-run -v
	* Improved bootstrap.sh

0.5.2 Duncan Coutts <duncan@haskell.org> Aug 2008
	* Suport building haddock documentaion
	* Self-reinstall now works on Windows
	* Allow adding symlinks to excutables into a separate bindir
	* New self-documenting config file
	* New install --reinstall flag
	* More helpful status messages in a couple places
	* Upload failures now report full text error message from the server
	* Support for local package repositories
	* New build logging and reporting
	* New command to upload build reports to (a compatible) server
	* Allow tilde in hackage server URIs
	* Internal code improvements
	* Many other minor improvements and bug fixes

0.5.1 Duncan Coutts <duncan@haskell.org> June 2008
	* Restore minimal hugs support in dependency resolver
	* Fix for disabled http proxies on Windows
	* Revert to global installs on Windows by default

0.5.0 Duncan Coutts <duncan@haskell.org> June 2008
	* New package dependency resolver, solving diamond dep problem
	* Integrate cabal-setup functionality
	* Integrate cabal-upload functionality
	* New cabal update and check commands
	* Improved behavior for install and upgrade commands
	* Full Windows support
	* New command line handling
	* Bash command line completion
	* Allow case insensitive package names on command line
	* New --dry-run flag for install, upgrade and fetch commands
	* New --root-cmd flag to allow installing as root
	* New --cabal-lib-version flag to select different Cabal lib versions
	* Support for HTTP proxies
	* Improved cabal list output
	* Build other non-dependent packages even when some fail
	* Report a summary of all build failures at the end
	* Partial support for hugs
	* Partial implementation of build reporting and logging
	* More consistent logging and verbosity
	* Significant internal code restructuring

0.4 Duncan Coutts <duncan@haskell.org> Oct 2007
	* Renamed executable from 'cabal-install' to 'cabal'
	* Partial Windows compatability
	* Do per-user installs by default
	* cabal install now installs the package in the current directory
	* Allow multiple remote servers
	* Use zlib lib and internal tar code and rather than external tar
	* Reorganised configuration files
	* Significant code restructuring
	* Cope with packages with conditional dependencies

0.3 and older versions by Lemmih, Paolo Martini and others 2006-2007
	* Switch from smart-server, dumb-client model to the reverse
	* New .tar.gz based index format
	* New remote and local package archive format


bootstrap.sh
#!/bin/sh

# A script to bootstrap cabal-install.

# It works by downloading and installing the Cabal, zlib and
# HTTP packages. It then installs cabal-install itself.
# It expects to be run inside the cabal-install directory.

# install settings, you can override these by setting environment vars
PREFIX=${PREFIX:-${HOME}/.cabal}
#VERBOSE
#EXTRA_CONFIGURE_OPTS

# programs, you can override these by setting environment vars
GHC=${GHC:-ghc}
GHC_PKG=${GHC_PKG:-ghc-pkg}
WGET=${WGET:-wget}
CURL=${CURL:-curl}
FETCH=${FETCH:-fetch}
TAR=${TAR:-tar}
GUNZIP=${GUNZIP:-gunzip}
SCOPE_OF_INSTALLATION="--user"


for arg in $*
do
  case "${arg}" in
    "--user")
      SCOPE_OF_INSTALLATION=${arg}
      shift;;
    "--global")
      SCOPE_OF_INSTALLATION=${arg}
      PREFIX="/usr/local"
      shift;;
    *)
      echo "Unknown argument or option, quitting: ${arg}"
      echo "usage: bootstrap.sh [OPTION]"
      echo
      echo "options:"
      echo "   --user    Install for the local user (default)"
      echo "   --global  Install systemwide (must be run as root)"
      exit;;
  esac
done


# Versions of the packages to install.
# The version regex says what existing installed versions are ok.
PARSEC_VER="3.1.1";    PARSEC_VER_REGEXP="[23]\."  # == 2.* || == 3.*
NETWORK_VER="2.3.0.2"; NETWORK_VER_REGEXP="2\."    # == 2.*
CABAL_VER="1.10.1.0";  CABAL_VER_REGEXP="1\.10\.[^0]"  # == 1.10.* && >= 1.10.1
TRANS_VER="0.2.2.0";   TRANS_VER_REGEXP="0\.2\."   # == 0.2.*
MTL_VER="2.0.1.0";     MTL_VER_REGEXP="[12]\."     # == 1.* || == 2.*
HTTP_VER="4000.1.1";   HTTP_VER_REGEXP="4000\.[01]\." # == 4000.0.* || 4000.1.*
ZLIB_VER="0.5.3.1";    ZLIB_VER_REGEXP="0\.[45]\." # == 0.4.* || ==0.5.*
TIME_VER="1.2.0.4"     TIME_VER_REGEXP="1\.[12]\." # == 0.1.* || ==0.2.*

HACKAGE_URL="http://hackage.haskell.org/packages/archive"

die () {
  echo
  echo "Error during cabal-install bootstrap:"
  echo $1 >&2
  exit 2
}

# Check we're in the right directory:
grep "cabal-install" ./cabal-install.cabal > /dev/null 2>&1 \
  || die "The bootstrap.sh script must be run in the cabal-install directory"

${GHC} --numeric-version > /dev/null \
  || die "${GHC} not found (or could not be run). If ghc is installed make sure it is on your PATH or set the GHC and GHC_PKG vars."
${GHC_PKG} --version     > /dev/null \
  || die "${GHC_PKG} not found."
GHC_VER=`${GHC} --numeric-version`
GHC_PKG_VER=`${GHC_PKG} --version | cut -d' ' -f 5`
[ ${GHC_VER} = ${GHC_PKG_VER} ] \
  || die "Version mismatch between ${GHC} and ${GHC_PKG} If you set the GHC variable then set GHC_PKG too"

# Cache the list of packages:
echo "Checking installed packages for ghc-${GHC_VER}..."
${GHC_PKG} list --global ${SCOPE_OF_INSTALLATION} > ghc-pkg.list \
  || die "running '${GHC_PKG} list' failed"

# Will we need to install this package, or is a suitable version installed?
need_pkg () {
  PKG=$1
  VER_MATCH=$2
  if grep " ${PKG}-${VER_MATCH}" ghc-pkg.list > /dev/null 2>&1
  then
    return 1;
  else
    return 0;
  fi
  #Note: we cannot use "! grep" here as Solaris 9 /bin/sh doesn't like it.
}

info_pkg () {
  PKG=$1
  VER=$2
  VER_MATCH=$3

  if need_pkg ${PKG} ${VER_MATCH}
  then
    echo "${PKG}-${VER} will be downloaded and installed."
  else
    echo "${PKG} is already installed and the version is ok."
  fi
}

fetch_pkg () {
  PKG=$1
  VER=$2

  URL=${HACKAGE_URL}/${PKG}/${VER}/${PKG}-${VER}.tar.gz
  if which ${CURL} > /dev/null
  then
    ${CURL} --fail -C - -O ${URL} || die "Failed to download ${PKG}."
  elif which ${WGET} > /dev/null
  then
    ${WGET} -c ${URL} || die "Failed to download ${PKG}."
  elif which ${FETCH} > /dev/null
 	then
 	  ${FETCH} ${URL} || die "Failed to download ${PKG}."
  else
    die "Failed to find a downloader. 'curl', 'wget' or 'fetch' is required."
  fi
  [ -f "${PKG}-${VER}.tar.gz" ] \
    || die "Downloading ${URL} did not create ${PKG}-${VER}.tar.gz"
}

unpack_pkg () {
  PKG=$1
  VER=$2

  rm -rf "${PKG}-${VER}.tar" "${PKG}-${VER}"/
  ${GUNZIP} -f "${PKG}-${VER}.tar.gz" \
    || die "Failed to gunzip ${PKG}-${VER}.tar.gz"
  ${TAR} -xf "${PKG}-${VER}.tar" \
    || die "Failed to untar ${PKG}-${VER}.tar.gz"
  [ -d "${PKG}-${VER}" ] \
    || die "Unpacking ${PKG}-${VER}.tar.gz did not create ${PKG}-${VER}/"
}

install_pkg () {
  PKG=$1

  [ -x Setup ] && ./Setup clean
  [ -f Setup ] && rm Setup

  ${GHC} --make Setup -o Setup \
    || die "Compiling the Setup script failed"
  [ -x Setup ] || die "The Setup script does not exist or cannot be run"

  ./Setup configure ${SCOPE_OF_INSTALLATION} "--prefix=${PREFIX}" \
    --with-compiler=${GHC} --with-hc-pkg=${GHC_PKG} \
    ${EXTRA_CONFIGURE_OPTS} ${VERBOSE} \
    || die "Configuring the ${PKG} package failed"

  ./Setup build ${VERBOSE} \
    || die "Building the ${PKG} package failed"

  ./Setup install ${VERBOSE} \
    || die "Installing the ${PKG} package failed"
}

do_pkg () {
  PKG=$1
  VER=$2
  VER_MATCH=$3

  if need_pkg ${PKG} ${VER_MATCH}
  then
    echo
    echo "Downloading ${PKG}-${VER}..."
    fetch_pkg ${PKG} ${VER}
    unpack_pkg ${PKG} ${VER}
    cd "${PKG}-${VER}"
    install_pkg ${PKG} ${VER}
    cd ..
  fi
}

# Actually do something!

info_pkg "Cabal"        ${CABAL_VER}   ${CABAL_VER_REGEXP}
info_pkg "transformers" ${TRANS_VER}   ${TRANS_VER_REGEXP}
info_pkg "mtl"          ${MTL_VER}     ${MTL_VER_REGEXP}
info_pkg "parsec"       ${PARSEC_VER}  ${PARSEC_VER_REGEXP}
info_pkg "network"      ${NETWORK_VER} ${NETWORK_VER_REGEXP}
info_pkg "time"         ${TIME_VER}    ${TIME_VER_REGEXP}
info_pkg "HTTP"         ${HTTP_VER}    ${HTTP_VER_REGEXP}
info_pkg "zlib"         ${ZLIB_VER}    ${ZLIB_VER_REGEXP}

do_pkg   "Cabal"        ${CABAL_VER}   ${CABAL_VER_REGEXP}
do_pkg   "transformers" ${TRANS_VER}   ${TRANS_VER_REGEXP}
do_pkg   "mtl"          ${MTL_VER}     ${MTL_VER_REGEXP}
do_pkg   "parsec"       ${PARSEC_VER}  ${PARSEC_VER_REGEXP}
do_pkg   "network"      ${NETWORK_VER} ${NETWORK_VER_REGEXP}
do_pkg   "time"         ${TIME_VER}    ${TIME_VER_REGEXP}
do_pkg   "HTTP"         ${HTTP_VER}    ${HTTP_VER_REGEXP}
do_pkg   "zlib"         ${ZLIB_VER}    ${ZLIB_VER_REGEXP}

install_pkg "cabal-install"

echo
echo "==========================================="
CABAL_BIN="$PREFIX/bin"
if [ -x "$CABAL_BIN/cabal" ]
then
    echo "The 'cabal' program has been installed in $CABAL_BIN/"
    echo "You should either add $CABAL_BIN to your PATH"
    echo "or copy the cabal program to a directory that is on your PATH."
    echo
    echo "The first thing to do is to get the latest list of packages with:"
    echo "  cabal update"
    echo "This will also create a default config file (if it does not already"
    echo "exist) at $HOME/.cabal/config"
    echo
    echo "By default cabal will install programs to $HOME/.cabal/bin"
    echo "If you do not want to add this directory to your PATH then you can"
    echo "change the setting in the config file, for example you could use:"
    echo "symlink-bindir: $HOME/bin"
else
    echo "Sorry, something went wrong."
    echo "The 'cabal' executable was not successfully installed into"
    echo "$CABAL_BIN/"
fi
echo

rm ghc-pkg.list
